### ISSUE TITLE
Query optimizer runtime is exponential in the number of subquery where clauses that reference the parent

### ISSUE THREAD

**Issue by kku1993** [duckdb/duckdb#6433]
### What happens?

In our system, we generate SQL queries with "stages", where each stage is a CTE that builds off of a previous stage. For example:

```
WITH stage0 AS (...),
stage1 AS (SELECT * from stage0),
stage2 AS (SELECT * from stage1),
...
```

We noticed that duckdb query optimizer's runtime is exponential in the number of "stages" that we have. Upon further investigation, we've identified WHERE clauses in subqueries that reference the parent as particularly problematic for the query optimizer.

For example (see repro section below for complete example):

```sql
WITH ref AS (...),
stage0 AS (...),
stage1 AS (SELECT stage0.*, (SELECT ... FROM ref WHERE ref.id = stage0.id) as foo from stage0),
stage2 AS (SELECT stage1.*, (SELECT ... FROM ref WHERE ref.id = stage1.id) as foo from stage1),
```

On my M1 mac, here's the amount of time it took to execute our test query with different number of stages

```
# Stages |   Time
# -------+-------
# 18     |  2.40s
# 19     |  4.34s
# 20     |  7.92s
# 21     |  15.2s
# 22     |  29.8s
# 23     |  59.4s
```

The query runs fine if we either:
1. Disable the optimizer OR
2. Remove subqueries that reference the parent

### To Reproduce

I've created a script to generate SQL that will trigger this behavior: https://gist.github.com/kku1993/ddb9bdbfc32bdb0fb0fb64e44dcd53cb

<details>
  <summary>Sample SQL with 25 stages</summary>
  
  ```sql
PRAGMA enable_optimizer;

with stage0 as (
  select 'b' as id, 1 as value
),
ref as (
  select 'a' as id, 10 as value
),
stage1 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage0 a
),
stage2 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage1 a
),
stage3 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage2 a
),
stage4 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage3 a
),
stage5 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage4 a
),
stage6 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage5 a
),
stage7 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage6 a
),
stage8 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage7 a
),
stage9 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage8 a
),
stage10 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage9 a
),
stage11 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage10 a
),
stage12 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage11 a
),
stage13 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage12 a
),
stage14 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage13 a
),
stage15 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage14 a
),
stage16 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage15 a
),
stage17 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage16 a
),
stage18 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage17 a
),
stage19 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage18 a
),
stage20 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage19 a
),
stage21 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage20 a
),
stage22 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage21 a
),
stage23 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage22 a
),
stage24 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage23 a
),
stage25 as (
  select
    a.*,
    (
      SELECT
        MIN("value") as value
      FROM (
        SELECT * FROM ref
        -- THIS APPEARS TO BE THE PROBLEM
        WHERE id = a.id
        -- WHERE id = 'a'
      )
    ) AS ref_value
  from stage24 a
),
output as (
  select * from stage25
)
select * from output

  ```
</details>

### OS:

macOS 13.2.1 (22D68), Apple M1 Pro

### DuckDB Version:

v0.7.0 f7827396d7

### DuckDB Client:

shell

### Full Name:

Kevin Ku

### Affiliation:

Watershed

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by l1t1** [comment]
my test data sql
```
select 'with t0 as(select 1 a)'s
union all
select ',t'||(n+1)||' as (select a from t'||(n)||')' from range(50)t(n)
union all
select 'select * from t50';
```
run no problem
```
with t0 as(select 1 a)
,t1 as (select a from t0)
,t2 as (select a from t1)
,t3 as (select a from t2)
,t4 as (select a from t3)
,t5 as (select a from t4)
,t6 as (select a from t5)
,t7 as (select a from t6)
,t8 as (select a from t7)
,t9 as (select a from t8)
,t10 as (select a from t9)
,t11 as (select a from t10)
,t12 as (select a from t11)
,t13 as (select a from t12)
,t14 as (select a from t13)
,t15 as (select a from t14)
,t16 as (select a from t15)
,t17 as (select a from t16)
,t18 as (select a from t17)
,t19 as (select a from t18)
,t20 as (select a from t19)
,t21 as (select a from t20)
,t22 as (select a from t21)
,t23 as (select a from t22)
,t24 as (select a from t23)
,t25 as (select a from t24)
,t26 as (select a from t25)
,t27 as (select a from t26)
,t28 as (select a from t27)
,t29 as (select a from t28)
,t30 as (select a from t29)
,t31 as (select a from t30)
,t32 as (select a from t31)
,t33 as (select a from t32)
,t34 as (select a from t33)
,t35 as (select a from t34)
,t36 as (select a from t35)
,t37 as (select a from t36)
,t38 as (select a from t37)
,t39 as (select a from t38)
,t40 as (select a from t39)
,t41 as (select a from t40)
,t42 as (select a from t41)
,t43 as (select a from t42)
,t44 as (select a from t43)
,t45 as (select a from t44)
,t46 as (select a from t45)
,t47 as (select a from t46)
,t48 as (select a from t47)
,t49 as (select a from t48)
,t50 as (select a from t49)
select * from t50
;
```

**Comment by kku1993** [comment]
@l1t1 please the the reproduction sql I posted in the `To Reproduce` section in the original issue report!

**Comment by l1t1** [comment]
the link you post is blocked. 
could you attach the file here.

**Comment by Mytherin** [comment]
Thanks for the report! I have pushed a fix in #6740
